stackslib/src/chainstate/burn/atc.rs

```diff
@@ -0,0 +1,1403 @@
+    /// (1) strictly less than X%, and (2) strictly less than any Pr[X% - c] for 0 < c < X.
+    /// * If the ATC carryover is less than 20%, the null miner has an overwhelmingly likely chance
+    /// of winning (>95%).
+    ///
+    /// The logistic curve fits the points (atc=0.2, null_prob=0.75) and (atc=0.8, null_prob=0.01).
+    fn null_miner_logistic(atc: f64) -> f64 {
+        // recall the inverted logistic function:
+        //
+        //                 L
+        // f(x) = ---------------------
+        //                -k * (x0 - x)
+        //           1 + e
+        //
+        // It is shaped like a *backwards* "S" -- it approaches L as `x` tends towards negative
+        // infinity, and it approaches 0 as `x` tends towards positive infinity.  This function is
+        // the null miner advantage function, where `x` is the ATC carryover value.
+        //
+        // We need to drive x0 and k from our two points:
+        //
+        // (x1, y1) = (0.2, 0.75)
+        // (x2, y2) = (0.8, 0.01)
+        //
+        // to derive L, x0, and k:
+        // L = 0.8
+        // z = ln(L/y1 - 1) / ln(L/y2 - 1)
+        // x0 = (x1 - z * x2) / (1 - z)
+        // k = ln(L/y1 - 1) / (x1 - x0)
+        //
+        // The values for x0 and k were generated with the following GNU bc script:
+        // ```
+        // $ cat /tmp/variables.bc
+        // scale=32
+        // supremum=0.8   /* this is L */
+        // x1=0.2
+        // y1=0.75
+        // x2=0.8
+        // y2=0.01
+        // z=l(supremum/y1 - 1)/l(supremum/y2 -1)
+        // x0=(x1 - z * x2)/(1 - z)
+        // k=l(supremum/y1 - 1)/(x1 - x0)
+        // print "x0 = "; x0
+        // print "k = "; k
+        // ```
+        //
+        // This script evaluates to:
+        // ```
+        // $ bc -l < /tmp/variables.bc
+        // x0 = .42957690816204645842320195118064
+        // k = 11.79583008928205260028158351938437
+        // ```
+
+        let L: f64 = 0.8;
+
+        // truncated f64
+        let x0: f64 = 0.42957690816204647;
+        let k: f64 = 11.795830089282052;
+
+        // natural logarithm constant
+        let e: f64 = 2.718281828459045;
+
+        let adv = L / (1.0 + e.powf(-k * (x0 - atc)));
+        adv
+    }
+
+    #[test]
+    fn make_null_miner_lookup_table() {
+        use crate::chainstate::burn::atc::ATC_LOOKUP;
+        let mut lookup_table = Vec::with_capacity(1024);
+        for atc in 0..1024 {
+            let fatc = (atc as f64) / 1024.0;
+            let lgst_fatc = null_miner_logistic(fatc);
+            let lgst_rational = AtcRational::from_f64_unit(lgst_fatc);
+            assert_eq!(ATC_LOOKUP[atc], lgst_rational);
+            lookup_table.push(lgst_rational);
+        }
+        println!("[");
+        for lt in lookup_table.into_iter() {
+            let inner = lt.into_inner();
+            println!("   AtcRational(Uint256({:?})),", &inner.0);
+        }
+        println!("]");
```

Comment:
Can you also test the other direction, i.e., compute the value of the logistic function using the lookup table and compare `result.to_f64()` to `null_miner_logistic(fatc)`?

Reply:
Sure